apps/aecore/src/aec_tx_pool_sync.erl

```diff
@@ -0,0 +1,477 @@
+              catch _:_ ->
+                  ok
+              end,
+              {ok, Done}
+          end,
+    do_local_action(Sync, finish, Fun, ?LOCAL_ACTION_TIMEOUT).
+
+%% To avoid locking up main process (and deadlock in the interaction with
+%% aec_peer_connection) we do all requests in a worker process. The worker
+%% is linked, and we also start a timer when spawning it. The timeout is
+%% 10s which should be sufficently more than the request timeout (7s) in
+%% aec_peer_connection.
+do_local_action(#sync{ id = Ref }, Action, ActionFun, Timeout) ->
+    Self = self(),
+    Fun = fun() ->
+              %% Receive the timeout message ref so
+              %% we can cancel it on success...
+              receive {go, SendRef} -> ok end,
+              Res =
+                  try
+                      ActionFun()
+                  catch _:Reason ->
+                      {error, Reason}
+                  end,
+              Self ! {local_action, Ref, SendRef, Action, Res}
+          end,
+    Pid        = spawn_link(Fun),
+    TimeoutMsg = {local_action, Ref, Pid, Action},
+    SendRef    = erlang:start_timer(Timeout, self(), TimeoutMsg),
+    Pid ! {go, SendRef}.
+
+kill_local_action(Worker, Ref, Action) ->
+    erlang:exit(Worker, kill),
+    %% Flush any late arriving result
+    receive {local_action, Ref, _, Action, _} -> ok
+    after 0 -> ok end.
+
+do_get(TxHashes) ->
+    lists:concat([ do_get_(Tx) || Tx <- TxHashes ]).
+
+do_get_(TxHash) ->
+    try
+        STx = aec_db:get_signed_tx(TxHash),
+        [STx]
+    catch _:_ ->
+        []
+    end.
+
+unfold({node, Path, Node}, Tree) ->
+    aeu_mp_trees:unfold(Path, Node, Tree);
+unfold({leaf, Path}, _Tree) ->
+    {leaf, Path};
+unfold({subtree, Path}, Tree) ->
+    get_subtree(Path, Tree);
+unfold({key, Key}, _Tree) ->
+    [{key, Key}].
+
+get_subtree(Key, _Tree) when bit_size(Key) =:= 32 * 8 ->
+    [{key, Key}];
+get_subtree(Path, Tree) ->
+    get_subtree(aeu_mp_trees:iterator_from(Path, Tree), Path, bit_size(Path), []).
+
+get_subtree(Iter, Path, S, Acc) ->
+    case aeu_mp_trees:iterator_next(Iter) of
+        {Key = <<Path:S/bits, _Rest/bits>>, _, NewIter} ->
+            get_subtree(NewIter, Path, S, [{key, Key} | Acc]);
+        _X ->
+            lists:reverse(Acc)
+    end.
+
+build_tx_mpt() ->
+    try
+        F = fun(TxHash, Tree) ->
+                aeu_mp_trees:put(TxHash, [], Tree)
+            end,
+        Tree = aec_db:fold_mempool(F, aeu_mp_trees:new()),
+        {ok, Tree}
+    catch _:R ->
+        {error, R}
+    end.
+
+-define(NODE_TAG, 0).
+-define(LEAF_TAG, 1).
+-define(SUBTREE_TAG, 2).
+-define(KEY_TAG, 3).
+
+serialize_unfolds(Us) ->
+    [ serialize_unfold(U) || U <- Us ].
+
+deserialize_unfolds(SUs) ->
+    [ deserialize_unfold(SU) || SU <- SUs ].
+
+serialize_unfold({node, Path, Node}) ->
+    aeu_rlp:encode(
+        aec_serialization:encode_fields(
+            [{type, int}, {path, binary}, {node, binary}],
```

Comment:
I see that `aec_peer_connection` calls `aec_blocks:serialize_to_binary/1` for serializing block that in turn relies on [`aec_object_serialization` recognizing internal tag `block` mapped to `100`](https://github.com/aeternity/epoch/blob/aa2017a6e94164ae19940565c29aea4b39a32b12/apps/aecore/src/aec_object_serialization.erl#L69) that is - correctly - not documented among [serializations meant to affect consensus](https://github.com/aeternity/protocol/blob/faf1815bcc8900a8c0020d09904b47cb9daf8640/serializations.md#block). Maybe these new serializations shall follow same path?